keras/models.py

```diff
@@ -93,95 +75,130 @@ def get_json_type(obj):
-                        param_dset = optimizer_weights_group.create_dataset(
-                            name,
-                            val.shape,
-                            dtype=val.dtype)
-                        if not val.shape:
-                            # scalar
-                            param_dset[()] = val
                         else:
-                            param_dset[:] = val
-        f.flush()
+                            name = 'param_' + str(i)
+                    else:
+                        if hasattr(w, 'name') and w.name:
+                            name = str(w.name)
+                        else:
+                            name = 'param_' + str(i)
+                    weight_names.append(name.encode('utf8'))
+                optimizer_weights_group.attrs['weight_names'] = weight_names
+                for name, val in zip(weight_names, weight_values):
+                    param_dset = optimizer_weights_group.create_dataset(
+                        name,
+                        val.shape,
+                        dtype=val.dtype)
+                    if not val.shape:
+                        # scalar
+                        param_dset[()] = val
+                    else:
+                        param_dset[:] = val
 
 
-def load_model(filepath, custom_objects=None, compile=True):
-    """Loads a model saved via `save_model`.
+def save_model(model, filepath_or_f, overwrite=True, include_optimizer=True):
+    """Save a model to a HDF5 file.
+
+    The saved model contains:
+        - the model's configuration (topology)
+        - the model's weights
+        - the model's optimizer's state (if any)
+
+    Thus the saved model can be reinstantiated in
+    the exact same state, without any of the code
+    used for model definition or training.
 
     # Arguments
-        filepath: String, path to the saved model.
+        model: Keras model instance to be saved.
+        filepath_or_f: either a String, path to the saved model, or an opened HDF file (or group of it)
+        overwrite: Whether we should overwrite any existing
+            model at the target location, or instead
+            ask the user with a manual prompt.
+        include_optimizer: If True, save optimizer's state together.
+
+    # Raises
+        ImportError: if h5py is not available.
+    """
+    if h5py is None:
+        raise ImportError('`save_model` requires h5py.')
+
+    if isinstance(filepath_or_f, str):
+        filepath = filepath_or_f
+
+        # If file exists and should not be overwritten.
+        if not overwrite and os.path.isfile(filepath):
+            proceed = ask_to_proceed_with_overwrite(filepath)
+            if not proceed:
+                return
+
+        with h5py.File(filepath, 'w') as f:
+            _save_model(model, f, include_optimizer)
+            f.flush()
+    else:
+        f = filepath_or_f
+        _save_model(model, f, include_optimizer)
```

Comment:
A flush is missing either here, or at the end of _save_model itself.

Reply:
I am not sure this works: if `f` is a H5 group, `flush` is not an attribute of `f`.